File: frontend/src/types/product.ts
/**
 * Types for Product management.
 *
 * Products are derived outputs from an imported project:
 * conversion, api, mcp, agents.
 */

// Product types matching backend ProductType enum
export type ProductType = "conversion" | "api" | "mcp" | "agents";

// Product status matching backend ProductStatus enum
export type ProductStatus =
  | "pending"
  | "in_progress"
  | "paused"
  | "completed"
  | "failed"
  | "unavailable";

// Product interface matching backend ProductResponse
export interface Product {
  id: string;
  project_id: string;
  project_name: string;
  product_type: ProductType;
  status: ProductStatus;
  workspace_path: string;
  session_id: string | null;
  output_directory: string | null;
  created_at: string;
  updated_at: string;
}

// List response from API
export interface ProductListResponse {
  products: Product[];
  total: number;
}

// Request to create a product
export interface CreateProductRequest {
  project_id: string;
  product_type: ProductType;
}

// Available product types (matches backend AVAILABLE_PRODUCT_TYPES)
export const AVAILABLE_PRODUCT_TYPES: ProductType[] = ["conversion"];

// Product status configuration for UI display
export const productStatusConfig: Record<
  ProductStatus,
  { label: string; color: string; icon: string }
> = {
  pending: { label: "Pendente", color: "yellow", icon: "clock" },
  in_progress: { label: "Em andamento", color: "blue", icon: "loader" },
  paused: { label: "Pausado", color: "orange", icon: "pause" },
  completed: { label: "Concluido", color: "green", icon: "check" },
  failed: { label: "Falhou", color: "red", icon: "x" },
  unavailable: { label: "Indisponivel", color: "gray", icon: "lock" },
};

// Product type configuration for UI display (card rendering)
export const productTypeConfig: Record<
  ProductType,
  { title: string; description: string; icon: string }
> = {
  conversion: {
    title: "Conversao FastAPI",
    description:
      "Converte o projeto WinDev para uma aplicacao FastAPI + Jinja2 moderna",
    icon: "code",
  },
  api: {
    title: "API REST",
    description:
      "Gera uma API REST documentada a partir das procedures do projeto",
    icon: "server",
  },
  mcp: {
    title: "Servidor MCP",
    description:
      "Cria um servidor MCP para integracao com Claude e outras IAs",
    icon: "cpu",
  },
  agents: {
    title: "Agentes IA",
    description:
      "Gera agentes especializados para automacao de tarefas do sistema",
    icon: "bot",
  },
};

File: src/wxcode/models/product.py
"""
Model de Produto derivado de um projeto.

Representa outputs de um projeto importado: conversao, api, mcp, agents.
"""

from datetime import datetime
from enum import Enum
from typing import Optional

from beanie import Document, Indexed, Link
from pydantic import Field

from wxcode.models.project import Project


class ProductType(str, Enum):
    """Tipos de produtos que podem ser gerados de um projeto."""
    CONVERSION = "conversion"
    API = "api"
    MCP = "mcp"
    AGENTS = "agents"


class ProductStatus(str, Enum):
    """Status do ciclo de vida de um produto."""
    PENDING = "pending"
    IN_PROGRESS = "in_progress"
    PAUSED = "paused"
    COMPLETED = "completed"
    FAILED = "failed"
    UNAVAILABLE = "unavailable"


class Product(Document):
    """
    Representa um produto derivado de um projeto WinDev.

    Produtos sao outputs de um projeto importado, como uma conversao
    para FastAPI, uma API REST, um servidor MCP, ou agentes.
    """

    # Relacionamento
    project_id: Link[Project] = Field(..., description="Projeto de origem")

    # Tipo e configuracao
    product_type: ProductType = Field(..., description="Tipo do produto")
    workspace_path: str = Field(..., description="Caminho do workspace do produto")

    # Status
    status: ProductStatus = Field(
        default=ProductStatus.PENDING,
        description="Status atual do produto"
    )

    # Sessao e output
    session_id: Optional[str] = Field(
        default=None,
        description="ID da sessao de processamento (Claude Code, etc.)"
    )
    output_directory: Optional[str] = Field(
        default=None,
        description="Diretorio de saida do produto gerado"
    )

    # Timestamps
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)
    started_at: Optional[datetime] = Field(
        default=None,
        description="Quando o processamento iniciou"
    )
    completed_at: Optional[datetime] = Field(
        default=None,
        description="Quando o processamento completou"
    )

    class Settings:
        name = "products"
        use_state_management = True
        indexes = [
            "project_id",
            "product_type",
            "status",
            [("project_id", 1), ("product_type", 1)],
        ]

    def __str__(self) -> str:
        return f"Product({self.product_type.value}, status={self.status.value})"
